### DISCUSSION TITLE
Same tls-group-id registered by several providers: Which implementation gets executed?

### DISCUSSION THREAD

**Discussion by baentsch** [openssl/openssl#25612] [Q&A] [answered]
Assume you load two providers, each registering the same [OSSL_CAPABILITY_TLS_GROUP_ID](https://docs.openssl.org/master/man7/provider-base/#capabilities) or "code point". Short of specifying explicitly a [property query](https://docs.openssl.org/master/man7/ossl-guide-libraries-introduction/#property-query-strings) (unusual to do on the command line anyway), is there any implicit or explicit logic deciding which of the two implementations "wins"/gets used?

The [algorithm fetching](https://docs.openssl.org/master/man7/ossl-guide-libcrypto-introduction/#algorithm-fetching) logic does not seem to help, as it does not seem to be the algorithm name deciding which provider is used but rather the code point as received in a handshake: This seems logical as the algorithm name (as passed in the [s_client -groups parameter](https://docs.openssl.org/master/man1/openssl-s_client/#synopsis) gets transformed into the code point during handshake and the latter then drives subsequent processing logic. As now two providers have registered the capability to handle this code point/group id, it seems uncertain which provider is chosen.

Is this indeed so? If so, would it not be advisable to add at least a warning about this to the documentation (e.g., strongly suggesting specification of a property query string)? Or would it even be better to output a warning message if more than one provider registers a TLS group id? Or even more drastically, prohibit a second registration of an already known code point?

**Comment by baentsch** [comment, marked as answer]
Somewhat answering my own question here's the statement from [the FIPS provider](https://docs.openssl.org/3.0/man7/fips_module/#loading-the-fips-module-at-the-same-time-as-other-providers):

> If no property query is specified, or more than one implementation matches the property query then it is undefined which implementation of a particular algorithm will be returned.

So, changing the question to: Would it be sensible to add the same warning to any provider-based algorithm registration documentation, e.g., at https://docs.openssl.org/master/man7/provider-base/#capabilities and at https://docs.openssl.org/master/man7/provider/#operations?

**Reply by t8m** [reply]
> So, changing the question to: Would it be sensible to add the same warning to any provider-based algorithm registration documentation, e.g., at https://docs.openssl.org/master/man7/provider-base/#capabilities and at https://docs.openssl.org/master/man7/provider/#operations?

Sure, that would be a good idea.

**Comment by t8m** [comment]
> If so, would it not be advisable to add at least a warning about this to the documentation (e.g., strongly suggesting specification of a property query string)? Or would it even be better to output a warning message if more than one provider registers a TLS group id? Or even more drastically, prohibit a second registration of an already known code point?

Warning about this in documentation. Yes, why not?

Warning during the registration callback. IMO no, and furthermore we do not really have any _warning_ facility.

Prohibit a second registration. Definitely no.

**Comment by paulidale** [comment]
We deliberately didn't define an ordering for the result of an ambiguous fetch which leaves open the option of changing things in the future.
